File: engine/models/workspace-subscription.ts
import { ServerDate } from "./server-date"

export interface WorkspaceSubscriptionServerDto {
  workspace_id: string
  status: string
  seats_purchased: number
  seats_used: number
  seats_available: number
  trial_ends_at?: string | null
  current_period_end?: string | null
  cancel_at_period_end: boolean
  is_trial_active: boolean
  is_read_only: boolean
  has_stripe_customer?: boolean
  has_subscription?: boolean
  billing_provider?: string
  campaign?: string | null
}

export interface WorkspaceSubscriptionClientDto {
  workspaceId: string
  status: string
  seatsPurchased: number
  seatsUsed: number
  seatsAvailable: number
  trialEndsAt?: string | null
  currentPeriodEnd?: string | null
  cancelAtPeriodEnd: boolean
  isTrialActive: boolean
  isReadOnly: boolean
  hasStripeCustomer: boolean
  hasSubscription: boolean
  billingProvider?: string
  campaign?: string | null
}

export class WorkspaceSubscription {
  readonly workspaceId: string
  readonly status: string
  readonly seatsPurchased: number
  readonly seatsUsed: number
  readonly seatsAvailable: number
  readonly trialEndsAt?: ServerDate
  readonly currentPeriodEnd?: ServerDate
  readonly cancelAtPeriodEnd: boolean
  readonly isTrialActive: boolean
  readonly isReadOnly: boolean
  readonly hasStripeCustomer: boolean
  readonly hasSubscription: boolean
  readonly billingProvider?: string
  readonly campaign?: string | null

  constructor(params: WorkspaceSubscriptionClientDto) {
    this.workspaceId = params.workspaceId
    this.status = params.status
    this.seatsPurchased = params.seatsPurchased
    this.seatsUsed = params.seatsUsed
    this.seatsAvailable = params.seatsAvailable
    this.cancelAtPeriodEnd = params.cancelAtPeriodEnd
    this.isTrialActive = params.isTrialActive
    this.isReadOnly = params.isReadOnly
    this.hasStripeCustomer = params.hasStripeCustomer
    this.hasSubscription = params.hasSubscription
    this.billingProvider = params.billingProvider
    this.campaign = params.campaign ?? null

    this.trialEndsAt = params.trialEndsAt ? new ServerDate(params.trialEndsAt) : undefined
    this.currentPeriodEnd = params.currentPeriodEnd ? new ServerDate(params.currentPeriodEnd) : undefined

    Object.freeze(this)
  }

  private toDto(): WorkspaceSubscriptionClientDto {
    return {
      workspaceId: this.workspaceId,
      status: this.status,
      seatsPurchased: this.seatsPurchased,
      seatsUsed: this.seatsUsed,
      seatsAvailable: this.seatsAvailable,
      trialEndsAt: this.trialEndsAt?.toString() ?? null,
      currentPeriodEnd: this.currentPeriodEnd?.toString() ?? null,
      cancelAtPeriodEnd: this.cancelAtPeriodEnd,
      isTrialActive: this.isTrialActive,
      isReadOnly: this.isReadOnly,
      hasStripeCustomer: this.hasStripeCustomer,
      hasSubscription: this.hasSubscription,
      billingProvider: this.billingProvider,
      campaign: this.campaign ?? null,
    }
  }

  public toClientDto(): WorkspaceSubscriptionClientDto {
    return this.toDto()
  }

  get hasSeatShortage(): boolean {
    if (this.isTrialActive) {
      return false
    }

    if (this.seatsPurchased <= 0) {
      return true
    }

    return this.seatsUsed > this.seatsPurchased
  }

  static fromServerDto(dto: WorkspaceSubscriptionServerDto): WorkspaceSubscription {
    return new WorkspaceSubscription({
      workspaceId: dto.workspace_id,
      status: dto.status,
      seatsPurchased: dto.seats_purchased,
      seatsUsed: dto.seats_used,
      seatsAvailable: dto.seats_available,
      trialEndsAt: dto.trial_ends_at ?? null,
      currentPeriodEnd: dto.current_period_end ?? null,
      cancelAtPeriodEnd: dto.cancel_at_period_end,
      isTrialActive: dto.is_trial_active,
      isReadOnly: dto.is_read_only,
      hasStripeCustomer: dto.has_stripe_customer ?? false,
      hasSubscription: dto.has_subscription ?? true,
      billingProvider: dto.billing_provider,
      campaign: dto.campaign ?? null,
    })
  }
}

File: server/models/workspace_subscription.go
package models

import "time"

// WorkspaceSubscriptionStatus mirrors Stripe's status vocabulary so downstream
// logic can treat the values predictably.
type WorkspaceSubscriptionStatus string

const (
	WorkspaceSubscriptionStatusTrialing   WorkspaceSubscriptionStatus = "trialing"
	WorkspaceSubscriptionStatusActive     WorkspaceSubscriptionStatus = "active"
	WorkspaceSubscriptionStatusPastDue    WorkspaceSubscriptionStatus = "past_due"
	WorkspaceSubscriptionStatusCanceled   WorkspaceSubscriptionStatus = "canceled"
	WorkspaceSubscriptionStatusIncomplete WorkspaceSubscriptionStatus = "incomplete"
	WorkspaceSubscriptionStatusUnpaid     WorkspaceSubscriptionStatus = "unpaid"
	WorkspaceSubscriptionStatusUnknown    WorkspaceSubscriptionStatus = "unknown"
)

// WorkspaceSubscriptionBillingProvider records which system issued the subscription.
type WorkspaceSubscriptionBillingProvider string

const (
	WorkspaceSubscriptionBillingProviderStripe     WorkspaceSubscriptionBillingProvider = "stripe"
	WorkspaceSubscriptionBillingProviderTests      WorkspaceSubscriptionBillingProvider = "tests"
	WorkspaceSubscriptionBillingProviderSelfHosted WorkspaceSubscriptionBillingProvider = "self_hosted"
)

// WorkspaceSubscription persists the billing state for a workspace.
type WorkspaceSubscription struct {
	WorkspaceID            string                               `json:"workspace_id" gorm:"primaryKey;type:uuid"`
	Status                 WorkspaceSubscriptionStatus          `json:"status" gorm:"type:text;not null"`
	SeatsPurchased         int                                  `json:"seats_purchased" gorm:"not null;default:0"`
	TrialEndsAt            *time.Time                           `json:"trial_ends_at"`
	CurrentPeriodEnd       *time.Time                           `json:"current_period_end"`
	CancelAtPeriodEnd      bool                                 `json:"cancel_at_period_end" gorm:"not null;default:false"`
	BillingProvider        WorkspaceSubscriptionBillingProvider `json:"billing_provider" gorm:"type:text;not null;default:stripe"`
	CreatedAt              time.Time                            `json:"created_at"`
	UpdatedAt              time.Time                            `json:"updated_at"`
	StripeSubscriptionInfo *StripeSubscriptionInfo              `json:"stripe_subscription_info,omitempty" gorm:"foreignKey:WorkspaceID;references:WorkspaceID"`
}

// WorkspaceSubscriptionSnapshot is returned to clients with derived details.
type WorkspaceSubscriptionSnapshot struct {
	WorkspaceID       string                               `json:"workspace_id"`
	Status            WorkspaceSubscriptionStatus          `json:"status"`
	BillingProvider   WorkspaceSubscriptionBillingProvider `json:"billing_provider"`
	Campaign          AcquisitionCampaign                  `json:"campaign,omitempty"`
	SeatsPurchased    int                                  `json:"seats_purchased"`
	SeatsUsed         int                                  `json:"seats_used"`
	SeatsAvailable    int                                  `json:"seats_available"`
	TrialEndsAt       *time.Time                           `json:"trial_ends_at,omitempty"`
	CurrentPeriodEnd  *time.Time                           `json:"current_period_end,omitempty"`
	CancelAtPeriodEnd bool                                 `json:"cancel_at_period_end"`
	IsTrialActive     bool                                 `json:"is_trial_active"`
	IsReadOnly        bool                                 `json:"is_read_only"`
	HasStripeCustomer bool                                 `json:"has_stripe_customer"`
	HasSubscription   bool                                 `json:"has_subscription"`
}

// computeSeatsAvailable ensures we never surface negative availability values.
func computeSeatsAvailable(purchased, used int) int {
	remaining := purchased - used
	if remaining < 0 {
		return 0
	}
	return remaining
}

// DetermineReadOnly encodes when a workspace should be locked for writes.
// TODO: Re-enable subscription checks when billing is ready for production.
func (ws WorkspaceSubscription) DetermineReadOnly(now time.Time, seatsUsed int, hasValidCredentials bool) bool {
	// Temporarily disabled: all workspaces have full access while building out the product.
	return false

	// Original billing enforcement logic (commented out for development):
	// if ws.TrialEndsAt != nil && now.Before(*ws.TrialEndsAt) {
	// 	// During trial, need valid credentials and seats
	// 	if !hasValidCredentials || ws.SeatsPurchased <= 0 {
	// 		return true
	// 	}
	// 	return false
	// }
	//
	// if ws.CancelAtPeriodEnd {
	// 	if ws.CurrentPeriodEnd == nil {
	// 		return true
	// 	}
	// 	if now.After(*ws.CurrentPeriodEnd) {
	// 		return true
	// 	}
	// }
	//
	// if ws.Status == WorkspaceSubscriptionStatusActive {
	// 	return ws.SeatsPurchased <= 0
	// }
	//
	// return true
}
